### ISSUE TITLE
ngcc: @ViewChildren predicate cannot be interpreted

### ISSUE THREAD

**Issue by kukjevov** [angular/angular#29286]
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/compiler-cli

### Is this a regression?

Dont know if this worked before.

### Description

When running `ngcc` for my own library i get error `@ViewChildren predicate cannot be interpreted` during compilation of **esm2015**

## 🔬 Minimal Reproduction

Here is repository with my sample package https://github.com/kukjevov/sample, there is compiled result that is used in sample https://github.com/kukjevov/ivy-code-size/tree/sample

To reproduce problem:
`git clone https://github.com/kukjevov/ivy-code-size sample-ivy`
`cd .\sample-ivy\`
`git checkout sample`
`npm install`
`git reset --hard`
`npm run ngcc`

## 🔥 Exception or Error
<pre><code>
Compiling @ng/sample - esm2015
@ViewChildren predicate cannot be interpreted

</code></pre>

## 🌍  Your Environment

**Angular Version:**
*8.0.0-beta.8*

**Anything else relevant?**
Windows 10
Version	10.0.17134 Build 17134
**node** v10.13.0
**npm** 5.3.0

**Comment by alexzuza** [comment]
The reason of why you're facing this kind of errors is that you don't follow Angular Package Format.
Angular team is currently testing Angular material library which is built with the respect of APF.

To be more precise:
1) Your es2015 bundle wasn't built with `"annotateForClosureCompiler": true,` option so instead of having:

```
OptGroupComponent.propDecorators = {
    optionsChildren: [{ type: ViewChildren, args: [OptionComponent,] }],
    text: [{ type: Input }]
};
```
in built artefacts you have something like:

```
__decorate([
    ViewChildren(OptionComponent),
    __metadata("design:type", QueryList)
], OptGroupComponent.prototype, "optionsChildren", void 0);
```
And ngtsc compiler can't extract ViewChildren from your piece of code currently. 
So first thing you can try is to add `"annotateForClosureCompiler": true,` option to `angularCompilerOptions` in tsconfig so that your code will look similar to angular material code.

2) Your es5 code is not es5 code because you need to downlevel generated code after ngc compilation.

**Comment by kukjevov** [comment]
Hi, i did some changes to make it working. One of the changes was missing `"annotateForClosureCompiler": true`, but I am not sure if it was necessary, but only that change did not helped. 

Problem was that metadata for components were not stored in `metadata.json`. Not completely sure why, but problem was in module which had to import components directly from containing files not from `index` import path. Dont understand why, maybe some circular dependencies.

**Comment by JoostK** [comment]
@alexzuza Actually, ngcc does already support TypeScript's emit format of decorators, next to tsickle's format. So, `annotateForClosureCompiler` does not necessarily have to be enabled.

One thing I noticed in this repo is that the `es5` dist folder does not contain ES5 code, but ES2015. I see these files are generated using `nnrh` in the package.json, however I don't know what tool that is.

**Comment by kukjevov** [comment]
Hmm interesting that you both saying that ES5 code is ES2015, but as far as i look at these `*.js` files i can see only ES5 code with `ES2015` module format, which is not problem because i use webpack which will handle them anyway, but rest is pure ES5. `tsconfig.json` which produces this directory has `"target": "es5"` so if it is not es5 then there is something wrong with typescript :).

But it does not matter. I managed to make it working. Interesting thing that i found out is that ngc which produces `metadata.json` in *flat* form does not work if you are not carefull with imports and exports.

Reason why it was not working, was that i had many so called *Barrel* files (index.ts) and obviously ngc has problem with them if it want to flatten them. Code was working without any problem, but i had different format of package and `ngfactory.js` files were part of dist bundle. I was able to run application in `JIT` or `AOT` mode without any problem. But then i wanted to test `ngcc` and i found out i need to change package format much more closer to APF to make it working. But it was unable to generate flat `metadata.json` because of these *Barrel* files. This is quite undocumented *feature* which was only problem here. 

But now it is working and i will have to revise all other packages to make them prepared for IVY :), which looks great but i miss some documentation. 

PS: `nnrh` is just my tool that handles `templateUrl` and replace them with `"template": require('')` and copies `*.html` files, which in fact i think can be replaced with `enableResourceInlining` angular compiler option. Im going to test it now :) 

Thx for replies anyway :)

**Comment by JoostK** [comment]
@kukjevov thanks for elaborating. Yes, barrel files are a known source of trouble.

I looked into your repo yet again, and your ES5 sources have become actual ES5 overnight 😄

Yesterday's commit:

https://github.com/kukjevov/sample/blob/7231f6b78b5e6803d6f1e2de56f28daa39d68e39/dist/es5/components/option/option.component.js

Notice the `let OptionComponent = class OptionComponent` syntax here.

After today's changes:

https://github.com/kukjevov/sample/blob/ea3febb31a013ba739efbb4480b1dbf95661b4f9/dist/es5/src/components/option/option.component.js

This one is indeed just ESM5, which is fine 👍

Great to hear you have been able to resolve it. Hopefully this issue can help others that experience the same issue.

**Comment by alexzuza** [comment]
> @alexzuza Actually, ngcc does already support TypeScript's emit format of decorators, next to tsickle's format. So, `annotateForClosureCompiler` does not necessarily have to be enabled.

Thank you for that information. 
I just looked at the place where the error happened then imagined how the generated ViewChild/ContentChild(-dren) code should look like. After that I checked @angular/material artefacts for similar code and realized that it is the work for tsickle to produce such code. So I tried to build the library with `annotateForClosureCompiler` and it indeed worked. But then I faced the issues with non ES5 code in `es5` folder.
